src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenForEachTests.cs

```diff
@@ -2365,5 +2365,774 @@ .locals init (T V_0)
+                    //         foreach (var i in new C())
+                    Diagnostic(ErrorCode.ERR_BadGetEnumerator, "new C()").WithArguments("C.Enumerator1", "C.GetEnumerator()").WithLocation(7, 27)
+                    );
+        }
+
+        [Fact]
+        public void TestPreferEnumeratorPatternFromIEnumerableInterfaceThanViaExtension()
+        {
+            var source = @"
+using System;
+using System.Collections;
+using System.Collections.Generic;
+public class C : IEnumerable<int>
+{
+    public static void Main()
+    {
+        foreach (var i in new C())
+        {
+            Console.Write(i);
+        }
+    }
+    
+    public sealed class Enumerator1 : IEnumerator<int>
+    {
+        object IEnumerator.Current => Current;
+        public int Current { get; private set; }
+
+        public bool MoveNext() => Current++ != 3;
+
+        public void Dispose() {}
+
+        public void Reset() { }
+    }
+
+    public sealed class Enumerator2
+    {
+        public int Current { get; private set; }
+        public bool MoveNext() => Current++ != 4;
+    }
+
+    IEnumerator<int> IEnumerable<int>.GetEnumerator() => new C.Enumerator1();
+    IEnumerator IEnumerable.GetEnumerator() => new C.Enumerator1();
+}
+
+public static class Extensions
+{
+    public static C.Enumerator2 GetEnumerator(this C self) => new C.Enumerator2();
+}";
+            CompileAndVerify(source, parseOptions: TestOptions.RegularPreview,
+                     expectedOutput: "123");
+        }
+
+        [Fact]
+        public void TestGetEnumeratorPatternViaAmbiguousExtensions()
+        {
+            var source = @"
+using System;
+public class C
+{
+    public static void Main()
+    {
+        foreach (var i in new C())
+        {
+            Console.Write(i);
+        }
+    }
+    public sealed class Enumerator
+    {
+        public int Current { get; private set; }
+        public bool MoveNext() => Current++ != 3;
+    }
+}
+public static class Extensions1
+{
+    public static C.Enumerator GetEnumerator(this C self) => new C.Enumerator();
+}
+public static class Extensions2
+{
+    public static C.Enumerator GetEnumerator(this C self) => new C.Enumerator();
+}";
+            CreateCompilation(source, parseOptions: TestOptions.RegularPreview)
+                .VerifyDiagnostics(
+                    // (7,27): warning CS0278: 'C' does not implement the 'collection' pattern. 'Extensions1.GetEnumerator(C)' is ambiguous with 'Extensions2.GetEnumerator(C)'.
+                    //         foreach (var i in new C())
+                    Diagnostic(ErrorCode.WRN_PatternIsAmbiguous, "new C()").WithArguments("C", "collection", "Extensions1.GetEnumerator(C)", "Extensions2.GetEnumerator(C)").WithLocation(7, 27),
+                    // (7,27): error CS1579: foreach statement cannot operate on variables of type 'C' because 'C' does not contain a public instance definition for 'GetEnumerator'
```

Comment:
We should suppress this error, and make the previous warning an error about ambiguous methods.

Reply:
> and make the previous warning an error about ambiguous methods.

What do you mean by this

Reply:
Oh I see. This is usually a warning since if it fails we can try the next step. Here there's no next step to try, so make an error instead.

This will give us more work to if we ever add an extra step after this :) But I guess we'll cross that bridge when we come to it.

Reply:
I've done this, but it adds signficant noise to the ForEachLoopBinder. I don't know if the benefits are worth it.